tf_quant_finance/black_scholes/brownian_bridge.py

```diff
@@ -0,0 +1,130 @@
+        (float32).
+    name: str. The name for the ops created by this function.
+      Default value: None which is mapped to the default name `option_price`.
+
+  Returns:
+      A `Tensor` of the same shape as the input data which is the probability
+      of not touching the upper and lower barrier.
+  """
+  with tf.name_scope(name or 'double_brownian_bridge'):
+    x_start = tf.convert_to_tensor(x_start, dtype=dtype, name='x_start')
+    dtype = x_start.dtype
+    variance = tf.convert_to_tensor(variance, dtype=dtype, name='variance')
+    x_end = tf.convert_to_tensor(x_end, dtype=dtype, name='x_end')
+    barrier_diff = upper_barrier - lower_barrier
+
+    def summed_func(k):
+      return tf.exp(-2 * k * barrier_diff * (
+              k * barrier_diff + (x_end - x_start)) / variance) \
+             - tf.exp(-2 * (k * barrier_diff + x_start - upper_barrier) * (
+              k * barrier_diff + (x_end - upper_barrier)) / variance)
+
+    return tf.reduce_sum(
+      tf.concat(
+        tf.map_fn(summed_func, tf.range(-n_cutoff, n_cutoff + 1, dtype=dtype)),
+        axis=0),
+      axis=0)
+
+
+def brownian_bridge_single(*,
+                           x_start,
+                           x_end,
+                           variance,
+                           barrier,
+                           dtype=None,
+                           name=None):
+  """ Probability of not touching the barrier for a Brownian Bridge starting
+  at `x_start`, ending at `x_end` and with variance `variance`,
+  assuming that `x_start` and `x_end` are the same side of the barrier (either
+  both above or both below)
+  This can be used in Monte Carlo pricing for adjusting probability of
+  touching the barrier from discrete case to continuous case.
+  Typically in practise, the tensors `x_start`, `x_end` and `variance` should be
+  bi-dimensional (with time steps and paths being the 2 dimensions).
+
+  ## References
+
+  [1] Emmanuel Gobet. Advanced Monte Carlo methods for barrier and related
+  exotic options.
+  https://papers.ssrn.com/sol3/papers.cfm?abstract_id=1265669
+
+  Args:
+    x_start: Real `Tensor` of any shape and dtype
+    x_end: A real `Tensor` of the same dtype and compatible shape as
+      `x_start`
+    variance: A real `Tensor` of the same dtype and compatible shape as
+      `x_start`
+    barrier: A scalar `Tensor`
+    dtype: Optional `tf.DType`. If supplied, the dtype to be used for conversion
+      of any supplied non-`Tensor` arguments to `Tensor`.
+      Default value: None which maps to the default dtype inferred by TensorFlow
+        (float32).
+    name: str. The name for the ops created by this function.
+      Default value: None which is mapped to the default name `option_price`.
+
+  Returns:
+      A `Tensor` of the same shape as the input data which is the probability
+      of not touching the barrier.
+  """
+  with tf.name_scope(name or 'single_brownian_bridge'):
```

Comment:
The default name should be the same as the function name unless there is a specific reason not to do so. Should we use `brownian_bridge_single `?